Optimized making requests when selecting users' apps

scorpion35 opened the pull request:
By calling `Application.all` once, the GET request is only executed once per user, when `selected_apps` is called. Thanks @janpio for the tip!

Branch: "master" | Commit Type: "Bug fix" | Changes tested: "yes"

---
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
**Context:**
- I am trying to get all users in a account and fetch the app names each user has access to. 

**Problem:**
- If a user has access to 10 apps (as an example), `10` GET requests were sent to Apple.
- And if there are 100 users/account, `100 * 10 = 1000` GET requests would be sent to Apple.

**If it fixes an open issue, please link to the issue here.**
- https://github.com/fastlane/fastlane/issues/14070
- This doesn't fix the issue completely, but optimizes **the number of GET requests to only one per user** (which is a huge win I think).

### Description
- Get all `Applications` before the loop. `all_apps = Application.all`
- Run `select` on the local `all_apps` variable.

**Please describe in detail how you tested your changes.**
- I have about over 600 users in all of our accounts combined. When I ran my script for all users, I have only seen ~600 GET requests.
- I have also noticed it's very fast now, even if the user has access to over 250 apps.

Commit:
Optimized making requests when selecting users' apps

By calling `Application.all` once, the GET request is only executed once per user, when `selected_apps` is called. Thanks @janpio for the tip!

Branch: "master" | Commit Type: "Bug fix" | Changes tested: "yes"

Commit:
Removed test line

Branch: "master" | Commit Type: "Bug fix" | Changes tested: "yes"

scorpion35 commented:
Pushed this from my work computer, probably that's why the CLA check is getting confused? I can confirm only I have pushed this fix!

scorpion35 commented:
Do you know how I can sign then CLA? Tried looking for the link, but wasn't able to find it.

joshdholtz commented on spaceship/lib/spaceship/tunes/member.rb:
```diff
@@ -56,10 +56,11 @@ def preferred_currency
+        all_apps = Application.all
         raw_data["userSoftwares"]["value"]["grantedSoftwareAdamIds"].each do |app_id|
-          parsed_apps << Application.find(app_id)
+          parsed_apps << all_apps.select { |app| !app.apple_id.nil? && app.apple_id == app_id }
```
```suggestion
          parsed_apps << all_apps.select { |app| app.apple_id == app_id }
```
You probably don't need to check for `nil` here ^ nil will never equal `app_id` so this will clean it up a bit

joshdholtz reviewed (changes requested):
One small nitpick on a nil check otherwise this looks 👍

Commit:
Update spaceship/lib/spaceship/tunes/member.rb

Sounds good. I am checking for nil on `apple_id` because I got bitten randomly couple of times.

Co-Authored-By: scorpion35 <anne.raghu@gmail.com>

janpio commented:
(Argh, I should actually read what @googlebot writes: As the Github account that was used to create the PR is not the same as the one that was used to push, it will not check for the CLA for either one. @joshdholtz will have to decide if we can merge this in this state, or if we need a force push with your GitHub account and then probably a new PR to fix this)

scorpion35 commented:
haha no worries. I will wait for the word!

janpio commented:
> Also, do you have any update on my PR, please? I am currently referencing my script to local fastlane changes, would have to move it to Jenkins build farm soon. Would appreciate if it's merged soon.

I can't influence when or if we get word on the CLA problem. Quickest would probably be to force push to the branch so it is committed/pushed by your Github account and then close this one and open a new PR where the bot will be able to connect you to the CLA properly - then I can just merge your PR.

scorpion35 committed:
Revert "Update spaceship/lib/spaceship/tunes/member.rb
"

This reverts commit 250cd6414f19eb5bb920e90c5eafc63d523cc190.

scorpion35 committed:
Revert "Removed test line"

This reverts commit c018d138b6a5df4abb54f9c6d19e3169ed975974.

scorpion35 committed:
Revert "Optimized making requests when selecting users' apps"

This reverts commit a2cba7f4defbd2129fb4db87f1944b7b27b1f2ef.

scorpion35 commented:
@janpio I have tried several things, but it seems Github is always coming back to these changes. Can this PR be merged with an approver this time, please? I will submit from my personal computer from next time.

@joshdholtz 🙏 Sorry for mentioning, just hoping for an approval so the PR can be merged :)

scorpion35 commented:
Alternatively if you want to submit from your account, I am ok with that too :)

joshdholtz commented:
@scorpion35 Looks like Google says its okay 👌 ` CLAs are signed, but unable to verify author consent ` and you consent so you are 👍 Doing a one last quick review and test then we will merge

joshdholtz commented:
@scorpion35 It actually looks like there are no changes in this PR anymore 🤔 

<img width="1066" alt="screen shot 2019-01-20 at 1 15 49 pm" src="https://user-images.githubusercontent.com/401294/51443888-8eaef600-1cb5-11e9-9e12-0015bece5045.png">

scorpion35 committed:
Revert "Revert "Optimized making requests when selecting users' apps""

This reverts commit 3b2fe95fb38a07530e3944fe8a75abe70fced6b3.

scorpion35 committed:
Revert "Revert "Removed test line""

This reverts commit e3773210a1a4d74e39698f8c503f5db84b45b5b1.

scorpion35 committed:
Revert "Revert "Update spaceship/lib/spaceship/tunes/member.rb
""

This reverts commit d7c155747a699dba0b39b55be882cec14da822a1.

janpio commented:
(I am sure @joshdholtz will take care of this now that he is aware. 

I looked at your 2 other PRs: 

The initial commits were not changed as I could see. If I were in your situation, I would have reset my branch to the first commit _before_ you committed anything with the wrong account, then committed the changes (that were kept in my working copy) again with the correct user - and finally force (!) pushed to my branch to get rid of the old commit and replace them with the new ones. That should automatically have updated this PR, but to get @googlebot to look at it again you still would have needed to create a new PR. 

But yep, that is definitely one of those git workflows that are not very common and stuff can go wrong. With the git CLI I would have no idea how to do it to be honest, but I work with Sourcetree, a git UI, long enough to be able to do this consistently now. Thank god ;)

Thanks for trying!)
